lib/pf/person.pm

```diff
@@ -375,6 +388,42 @@ sub person_violations {
     return db_data(PERSON, $person_statements, 'person_violations_sql', $pid);
 }
 
+=head2 persons_without_nodes
+
+Get all the persons who are not the owner of at least one node.
+
+=cut
+
+sub persons_without_nodes {
+   return db_data(PERSON, $person_statements, 'persons_without_nodes_sql'); 
+}
+
+=head2 person_cleanup
+
+Clean all persons that are not the owner of a node and that are not a local account that is still valid
+
+=cut
+
+sub person_cleanup {
+    my @to_delete = map { $_->{pid} } persons_without_nodes();
+    my $now = DateTime->now();
+    foreach my $pid (@to_delete) {
+        my $password = pf::password::view($pid);
+        if(defined($password)){
+            my $expiration = DateTime::Format::MySQL->parse_datetime($password->{expiration});
+            my $cmp = DateTime->compare($now, $expiration);
+            if($cmp < 0){
+                get_logger->debug("Not deleting $pid because the local account is still valid.");
+                next;
+            }
+            # We delete the password too
+            pf::password::_delete($pid);
+        }
+        # We're all good for deletion
+        person_delete($pid);
+    }
+}
+
```

Comment:
As I thought about it since we are not doing bulk deletes in sql batching is not needed in these cases.
So this will not be problem for node_cleanup or person_cleanup.
However it will be slower since we are doing it in code instead of sql.
But like you said these are never really big tables not that big of a deal.

However there is no grace period for deletion from the password table.
This is a deviation from all the other expiration patterns we have.